src/lib/datatypes/include/sol-memdesc.h

```diff
@@ -0,0 +1,1089 @@
+     * the traditional use of struct sol_memdesc::defcontent.
+     *
+     * Should return 0 on success, negative errno on errors.
+     *
+     * @see sol_memdesc_init_defaults()
+     */
+    int (*init_defaults)(const struct sol_memdesc *desc, void *memory);
+    /**
+     * @brief sets the content of a memory.
+     *
+     * If provided, will be used to set the memory instead of the
+     * traditional code that will, for example, strdup() and free()
+     * strings.
+     *
+     * The parameter @c ptr_content is a pointer to the actual content,
+     * depends on the actual type. If a SOL_MEMDESC_TYPE_BOOLEAN, for
+     * example, it must be a @c bool*.
+     *
+     * Should return 0 on success, negative errno on errors.
+     *
+     * @see sol_memdesc_set_content()
+     */
+    int (*set_content)(const struct sol_memdesc *desc, void *memory, const void *ptr_content);
+    /**
+     * @brief copy the content from another memory.
+     *
+     * If provided, will be used to set the memory instead of the
+     * traditional code that will, for example, strdup() and free()
+     * strings.
+     *
+     * Should return 0 on success, negative errno on errors.
+     *
+     * @see sol_memdesc_copy()
+     */
+    int (*copy)(const struct sol_memdesc *desc, const void *src_memory, void *dst_memory);
+    /**
+     * @brief compare the content of two memories.
+     *
+     * If provided, will be used to compare the memory contents
+     * instead of the traditional code that will, for example,
+     * call strcmp() on strings.
+     *
+     * Should return 0 for equal, <0 if a_memory is smaller, >0 if b_memory is smaller.
+     * On error, return 0 and set errno.
+     *
+     * @see sol_memdesc_compare()
+     */
+    int (*compare)(const struct sol_memdesc *desc, const void *a_memory, const void *b_memory);
+    /**
+     * @brief free the contents (internal memory) of a memory.
+     *
+     * If provided, will be used to free the contents of a memory
+     * instead of the traditional code that will, for example, free()
+     * strings.
+     *
+     * Should return 0 on success, negative errno on errors.
+     *
+     * @see sol_memdesc_free_content()
+     */
+    int (*free_content)(const struct sol_memdesc *desc, void *memory);
+    union {
+        const struct sol_memdesc_ops_array *array;
+    };
+};
+
+/**
+ * @brief Data type to describe a memory region.
+ */
+struct sol_memdesc {
+#ifndef SOL_NO_API_VERSION
+#define SOL_MEMDESC_API_VERSION (1) /**< @brief API version to use in struct sol_memdesc::api_version */
+    uint16_t api_version; /**< @brief API version, must match SOL_MEMDESC_API_VERSION at runtime */
+#endif
+    /**
+     * @brief size in bytes of the member memory.
+     *
+     * Usually this is @c sizeof(type), if a structure it will account
+     * for all members plus paddings.
+     *
+     * This is only used for SOL_MEMDESC_TYPE_STRUCTURE and
+     * SOL_MEMDESC_TYPE_ARRAY.
+     */
+    uint16_t size;
```

Comment:
Why this should be used only for `SOL_MEMDESC_TYPE_STRUCTURE` and `SOL_MEMDESC_TYPE_STRUCTURE`? Wouldn't be easier ask for size when defining each `sol_memdesc`? Or are you trying to avoid mistakes on this use? I'm afraid that having this size to be used sometimes may lead to more confusion than explicitly asking for it all the time....